### ISSUE TITLE
Questions about the SG-AD

### ISSUE THREAD

**Issue by llCurious** [ege-erdogan/splitguard#2]
Hey @ege-erdogan , i noticed your work and have some questions as follows:
- In my understanding, the training data (feature + label) are vertically splited in split learning. While in the paper (same in unleashing the tiger), it looks like federeated learning, where each client hold entire features and labels for different training samples.
- In SG-AD, if the gradient list used for train LOF is constructed as: 1) run batch training locally; 2) append the gradients for each layer (or gradients for some specific layer like the last layer?) to the list. Therefore the gradient list is of shape (#num_of_iterations, K, W, H), where K denotes the number of model layers and W, H denote the weight shape for some layer.

**Comment by ege-erdogan** [comment]
Hey @llCurious, thanks for the questions!

* Split learning can be used in both of the setups you described. We mainly focused on the scenario where all clients have all features also as in the FSHA paper but some other work considers the vertical scenario with clients storing different sets of features. At a first glance, I don't think there is much difference between both scenarios when you target a single client with FSHA, so I'd say our methods applicable but we'd need specific experiments to confirm that, especially for SG-AD to show that collecting local grads with a subset of features while the actual server model is trained with all features does not break things. There shouldn't be a difference for SG-LC since it is only concerned with the labels' effect on the learning process. 

* Yes, that's right, although we flatten the each gradient into 1D, so if the local model is trained with N batches we end up with a list of N one-dim tensors. 

Hope these clear things up. Let me know if you have further questions.

**Comment by llCurious** [comment]
Thx for your quick reply!
-  As to the first question, it seems to be the attacks (client-side attack) in FSHA and SG-LC require the client to know labels, so as to launch attacks or defenses. If the scenario is vertical SL, then maybe the label is held in only one client (with the others only providing features)
- Yep. I get it. You just flatten the gradients in the model (all the layers) and stack them to get (N, D) gradient list, where D = \sum{W_i * H_i}.

**Comment by ege-erdogan** [comment]
Yes that's right. Clients need to have the labels to run SG-LC, so in the vertical setup only the clients with the labels could run SG-LC.
